src/borg/helpers/fs.py

```diff
@@ -19,6 +21,37 @@
 py_37_plus = sys.version_info >= (3, 7)
 
 
+def ensure_dir(path, name=None, mode=0o777, pretty_deadly=True):
+    """
+    Ensures that the dir exists with the right permissions.
+    1) Make sure the directory exists in a race-free operation
+    2) If mode is not None and the directory has been created, give the right
+    permissions
+    3) If pretty_deadly is True, catch fatal exceptions, reraise them with a pretty
+    message.
+    Returns True if the file has been created and has the right permissions,
+    False otherwise. If a deadly exception happened it is reraised.
+    """
+    try:
+        os.makedirs(path, mode=mode, exist_ok=True)
+    except FileExistsError:
+        if pretty_deadly:
+            raise Error(e.args[1])
+        else:
+            raise
+    except FileNotFoundError as e:
+        if pretty_deadly:
+            raise Error(e.args[1])
+        else:
+            raise
+    except OSError as e:
+        if pretty_deadly:
+            raise Error(e.args[1])
+        else:
+            raise
+    return True
```

Comment:
seen this?

Reply:
btw, the current code is a mixup of 2 fundamentally different ways to deal with error handling:

- C / Go style: return some 0 / 1 / negative / positive / True / False value to signal success / failure. In the worse case, this is even in-band signalling (e.g. most values meaning a measured value and some specific values, like 0, meaning an error).

This is bad and leads to unpretty code structure (tons of if/else and bubbling up of error codes, bugs if the "if" is "forgotten") and in case of in-band signalling also to other problems.

- Python / Java(?) style: do **not** return error codes, but throw meaningful exceptions. If there is e.g. some measurement value, return that (if the measurement succeeded). Exceptions can be catched at any level in the code (or not, if fatal).